src/core/commit.rs

```diff
@@ -601,142 +601,66 @@ impl<J: Jet> CommitNode<J> {
-                                &wit_arrow,
-                            );
-
-                            // Since they have no children we can do the entire second pass here...
-                            let wit_imr =
-                                Imr::compute_pass2(wit_first_pass_imr, &x.0.inner, &wit_arrow);
-                            let amr =
-                                Amr::compute(&x.0.inner, None, None, Some(&wit_val), &wit_arrow);
-                            let bounds = analysis::compute_bounds(x.0, None, None, &wit_arrow);
-                            let node = RedeemNode {
-                                inner: RedeemNodeInner::Witness(wit_val),
-                                cmr: x.0.cmr,
-                                imr: wit_imr,
-                                amr,
-                                ty: wit_arrow,
-                                bounds,
-                            };
-                            // ...and store the finalized node
-                            to_finalized.insert(wit_first_pass_imr, Rc::new(node));
-
-                            Ok(Some(wit_first_pass_imr))
-                        } else {
-                            Ok(Some(
-                                *first_pass
-                                    .get(&x)
-                                    .expect("Children come before parent in post order"),
-                            ))
-                        }
-                    }
-                    Some(x) => Ok(Some(
-                        *first_pass
-                            .get(&x)
-                            .expect("Children come before parent in post order"),
-                    )),
-                    None => Ok(None),
-                }
-            }
-
-            let left_imr = first_pass_child_imr(
-                unshare_witnesses,
-                commit.get_left(),
-                &mut witness,
-                &first_pass,
-                &mut to_finalized,
-            )?;
-            let right_imr = first_pass_child_imr(
-                unshare_witnesses,
-                commit.get_right(),
-                &mut witness,
-                &first_pass,
-                &mut to_finalized,
-            )?;
-
-            let final_ty = commit.0.arrow.finalize()?;
-            let value = if let CommitNodeInner::Witness = commit.0.inner {
-                let v = witness.next(&final_ty.target)?;
-                Some(v)
+        // Iterate over either the internally-shared DAG, or the internally-shared DAG
+        // with witnesses unshared. The borrowck makes this a little irritating to express.
+        let mut share = InternalSharing::default();
+        let mut unshare = dag::UnshareWitnessDisconnect::new(share.clone());
+        let tracker: &mut dyn dag::SharingTracker<&CommitNode<J>> = if unshare_witnesses {
+            &mut unshare
+        } else {
+            &mut share
+        };
```

Comment:
If the trait object is annoying, we could write a private method `finalize_helper<S>(&self, mut witness: W)` with `S` being the sharing tracker type, and wrap this in the public `finalize()` method.

Reply:
Ah, that's not too bad an idea. Let's keep it in mind the next time we're messing with this code.